### PR TITLE
Fixes #4789 Re-enable cache of feeds

### LINKED ISSUE #4789: Feeds caching isn't working
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version ✔ `3.10.9`
- Used the search feature to ensure that the bug hasn’t been reported before ✅

**Describe the bug**

Caching the feeds isn't working at the moment for two reasons:
1. The [WP Rocket Cache Feed](https://github.com/wp-media/wp-rocket-helpers/tree/master/cache/wp-rocket-cache-feed) helper plugin had the wrong feed URI pattern (latest change on [Sep 4, 2020](https://github.com/wp-media/wp-rocket-helpers/commit/03cbf5834f47e7428a6572f0801f707bbd39b9d0)). 
In the meantime, we changed the pattern to fix #2852. I've updated the helper plugin to match this.
2. Even with the fixed helper, caching won't take place because of the check for a `</html>` here:
https://github.com/wp-media/wp-rocket/blob/d7c3f04dc217b2c98f5cd16a9da119406c68c517/inc/classes/Buffer/class-tests.php#L355-L365

**To Reproduce**

1. Install the updated WP Rocket Cache Feed helper plugin.
2. Visit the URL of the feed.
4. Check the cache folder for the feed folder. It won't be there.

**Expected behavior**

Feeds caching should work when the respective helper is installed/activated.

**Additional context**

We can fix this by adding a new `is_feed_uri()` method in `WP_Rocket\Buffer\Tests`:
```PHP
function is_feed_uri(){
		global $wp_rewrite;
		$feed_uri = '/(?:.+/)?' . $wp_rewrite->feed_base . '(?:/(?:.+/?)?)?$';
		return (bool) preg_match( '#^(' . $feed_uri . ')$#i', $this->get_clean_request_uri() );
}
```

and use it when checking for the `</html>` element like this:
```PHP
if (
			$this->has_test( 'is_html' )
			&&
			! $this->is_html( $buffer )
			&&
			! defined( 'REST_REQUEST' )
			&&
			! $this->is_feed_uri()
		) {
			// Don't process if there isn't a closing </html>.
			$this->set_error( 'No closing </html> was found.' );
			return false;
		}
```
@piotrbak Maybe this should be tackled in the same release with #4133.

**Backlog Grooming (for WP Media dev team use only)**
- [x] Reproduce the problem
- [x] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description
This PR fix a problem concerning feeds caching.
For that the logic inside the Buffer\Tests class have been a bit changed to check if the uri is a feed before checking if the page is HTML.
This preventing the bug excluding feeds from cache.

Fixes #4789 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

No

## How Has This Been Tested?

- [x] Automated tests
- [x] Manually in local

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by CrochetFeve0251** [comment]
@Tabrisrp it looks good now

File: tests/Fixtures/inc/classes/Buffer/Tests/canProcessBuffer.php [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		],
+		'expected' => [
+			'buffer_results' => true,
+			'error' => [
+				'message' => '',
+				'data' => [],
+			],
+		],
+	],
+	'testShouldReturnTrueWhenIsFlux' => [
```
Comment (remyperona) [inline tests/Fixtures/inc/classes/Buffer/Tests/canProcessBuffer.php]:
```suggestion
	'testShouldReturnTrueWhenIsFeed' => [
```

File: tests/Unit/inc/classes/Buffer/Tests/isFeedUri.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+
+	/**
+	 * @dataProvider configTestData
+	 */
+	public function testShouldReturnAsExpected($config, $expected)
+	{
+		$wp_rewrite = new WP_Rewrite();
+		$wp_rewrite->feed_base = $config['feed_base'];
+		$GLOBALS['wp_rewrite'] = $wp_rewrite;
```
Comment (remyperona) [inline tests/Unit/inc/classes/Buffer/Tests/isFeedUri.php:40]: I would recommend to reset/unset the global in the `tearDown()` fixture, to make sure it's properly cleaned between each test.

File: tests/Unit/inc/classes/Buffer/Tests/isFeedUri.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		$this->tests->expects()->get_clean_request_uri()->andReturn($config['clean_uri']);
+		$this->assertEquals($expected, $this->tests->is_feed_uri());
```
Comment (remyperona) [inline tests/Unit/inc/classes/Buffer/Tests/isFeedUri.php]:
```suggestion
		$this->assertSame($expected, $this->tests->is_feed_uri());
```

File: tests/Unit/inc/classes/Buffer/Tests/canProcessBuffer.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	 * @dataProvider configTestData
+	 */
+	public function testShouldReturnAsExpected($config, $expected) {
+		$this->configureRocketFunction($config);
+		$this->configureHttpResponse($config);
+		$this->configureCache($config);
+		$this->configure404($config);
+		$this->configureSearch($config);
+		$this->configureHTML($config);
+		$this->assertEquals($expected['buffer_results'], $this->tests->can_process_buffer($config['buffer']));
```
Comment (remyperona) [inline tests/Unit/inc/classes/Buffer/Tests/canProcessBuffer.php]:
```suggestion
		$this->assertSame($expected['buffer_results'], $this->tests->can_process_buffer($config['buffer']));
```

File: tests/Unit/inc/classes/Buffer/Tests/canProcessBuffer.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		$this->assertEquals($expected['error'], $this->tests->get_last_error());
```
Comment (remyperona) [inline tests/Unit/inc/classes/Buffer/Tests/canProcessBuffer.php]:
```suggestion
		$this->assertSame($expected['error'], $this->tests->get_last_error());
```

**Review by remyperona — CHANGES_REQUESTED** [review]

File: tests/Unit/inc/classes/Buffer/Tests/isFeedUri.php (hunk 2 above)
Comment (remyperona) [reply tests/Unit/inc/classes/Buffer/Tests/isFeedUri.php:40]: This is still missing

File: tests/Unit/inc/classes/Buffer/Tests/canProcessBuffer.php (hunk 5 above)
Comment (remyperona) [reply tests/Unit/inc/classes/Buffer/Tests/canProcessBuffer.php]: ☝🏼

Comment (CrochetFeve0251) [reply tests/Unit/inc/classes/Buffer/Tests/canProcessBuffer.php]: @Tabrisrp ok changed it

File: tests/Unit/inc/classes/Buffer/Tests/isFeedUri.php (hunk 2 above)
Comment (CrochetFeve0251) [reply tests/Unit/inc/classes/Buffer/Tests/isFeedUri.php:40]: @Tabrisrp changed

**Review by remyperona — APPROVED** [review]

**Comment by vmanthos** [comment]
This is working as expected. 👍

Note: I also tested caching of REST API requests since the change in the code might have affected that as well.

**Review by vmanthos — APPROVED** [review]

**Pull request merged**